testing/test_optimizers.py

```diff
@@ -298,6 +288,62 @@ def test_mcmc(self):
         self.assertTrue(np.allclose(result.x, 0), msg="Optimizer failed to find optimum")
         self.assertTrue(np.allclose(result.fun, 0), msg="Incorrect function value returned")
 
+    def test_callback(self):
+        class DummyCallback(object):
+            def __init__(self):
+                self.counter = 0
+
+            def __call__(self, models):
+                self.counter += 1
```

Comment:
I think the model contains all  the data you need to accomplish something. I believe X and Y can even be updated in this callback as long as the model supports it (all models in GPflow do). 

If at some point some information is really missing, this can be added.